### PR TITLE
Moving the Utf8Formatter and Utf8Parser into S.P.Corelib

### PR DESCRIPTION
**Pull request by tannergooding**
As per #20873 (comment), I split the PR into multiple.

This is the third PR and moves the Utf8 Formatter and Parser into S.P.Corelib.
* As part of the update, it finishes fixing the Utf8Parser for Single/Double to match the new Utf16Parser changes.

### CODE REVIEW

**Comment by tannergooding** [comment]
CC. @stephentoub, @GrabYourPitchforks, @jkotas

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs [hunk 1]
```diff
@@ -66,34 +72,52 @@ private static bool TryParseNumber(ReadOnlySpan<byte> source, ref NumberBuffer n
...
+                {
+                    // For decimal and binary floating-point numbers, we only
+                    // need to store digits up to maxDigCount. However, we still
+                    // need to keep track of whether any additional digits past
+                    // maxDigCount were non-zero, as that can impact rounding
+                    // for an input that falls evenly between two representable
+                    // results.
+
+                    number.HasNonZeroTail = (c != '0');
+                }
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs]: Why is this change necessary as part of moving this into corelib?  Does this have any impact on throughput, seeing as though we're adding additional logic in what appears to be a few tight loops?

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs]: CoreFX tests won't pass otherwise.

The Utf8Parser is using the NumberBuffer implementation that was already in S.P.Corelib now. The `Digit` buffer it contains is dynamically sized to be exactly the right size (which is enough digits to hold the largest exact string + 1 digit for rounding and + 1 digit for the null terminator). It also tracks whether any digits that were specified past the rounding digit where zero, which allows it to be correctly rounded in all scenarios.

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs]: I'll check perf here, but I would expect some minimal impact (most of which should hopefully be handled by the branch predictor).

**Comment by tannergooding** [comment]
@stephentoub, how do we want to handle the mirror?

Merging this will cause the changes to mirror to CoreFX, which will cause `System.Memory` to no longer build.
The correct fix is to remove the code from System.Memory and ensure the type forward to S.P.Corelib exists, but that can't be done until after the CoreCLR package update happens.

I think the options are:
* Wait and do the mirror and package update together, or...
* For the mirror, add an additional commit which removes the existing NumberBuffer from System.Memory and replaces it with a link to the shared NumberBuffer implementation from corelib
  Then, remove all links and add the type-forward when the package update occurs

**Comment by jkotas** [comment]
Add temporary ifdefs around the offending code

**Comment by tannergooding** [comment]
> Add temporary ifdefs around the offending code

I think there is too much code that was changed to do that, it would basically be putting an `#ifdef` around all changes in this PR.

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs [hunk 2]
```diff
@@ -128,7 +171,7 @@ private static bool TryParseNumber(ReadOnlySpan<byte> source, ref NumberBuffer n
                     }
                 }
 
-                int numDigitsAfterDecimalToCopy = Math.Min(srcIndex - startIndexOfDigitsAfterDecimalToCopy, NumberBuffer.BufferSize - dstIndex - 1);
+                int numDigitsAfterDecimalToCopy = Math.Min(srcIndex - startIndexOfDigitsAfterDecimalToCopy, number.Digits.Length - 1);
```
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs]: need to fix this, it should be `number.Digits.Length - dstIndex - 1` (which can be `maxDigitCount - dstIndex`)

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs (hunk 1 above)
Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs]: @stephentoub, we have no perf tests that currently exercise this code (which is decimal, double, and float). We only have tests that cover the integer types.
* Working on getting some basic tests to cover this now...

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs]: Really basic perf tests (just mirroring the utf16 ones) here: https://github.com/dotnet/corefx/pull/33427

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs]: @stephentoub, @GrabYourPitchforks, perf numbers look similar to the Utf16Parser/Utf16Formatter differences.

For integers (both parsing and formatting), we are the essentially the same (some faster, some slower; all look within std deviation).
For floating-point (decimal, single, and double), we are essentially the same for formatting. For parsing, we essentially match the results that were in the PR changing the UTF16Parser
* For floating-point, we are faster on numbers that are <= 15 digits and with exponents <= 22. For those outside the range, we are ~5x slower but are now returning the correct result.
  * As mentioned on the Utf16Parser issue, we can improve this further for numbers <= 19 digits and with exponents <= 27 by using 80-bit extended precision multiplication/division
    We can improve this for numbers <= 34 digits and with exponents <= 38 by using 128-bit floating-point multiplication/division

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs]: Can't seem to upload the CSV files right now, GitHub is having issues doing so.

**Comment by stephentoub** [comment]
> I think the options are

It shouldn't (famous last words) be a long delay between this being merged and a new build being produced and consumed into corefx.  You could just wait to merge the mirror PR until that's happened, at which point you can add a commit with the appropriate fix-ups to the mirror PR.  Reasonable?

**Comment by stephentoub** [comment]
Ah, then yeah, if ifdef'ing doesn't work, that seems reasonable.

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs [hunk 3]
```diff
@@ -242,5 +242,15 @@ public static uint DivMod(uint numerator, uint denominator, out uint modulo)
         }
 
         #endregion Math Helper methods
+
+        //
+        // Enable use of ThrowHelper from TryFormat() routines without introducing dozens of non-code-coveraged "bytesWritten = 0; return false" boilerplate.
+        //
+        public static bool TryFormatThrowFormatException(out int bytesWritten)
+        {
+            bytesWritten = 0;
```
Comment (GrabYourPitchforks) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs]: Nit: swap this line with the line immediately below. Otherwise the write has an observable side effect that the JIT cannot elide.
(Don't let this block checkin.)

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/ParserHelpers.cs [hunk 4]
```diff
@@ -51,5 +51,24 @@ public static bool IsDigit(int i)
         {
             return (uint)(i - '0') <= ('9' - '0');
         }
+
+        //
+        // Enable use of ThrowHelper from TryParse() routines without introducing dozens of non-code-coveraged "value= default; bytesConsumed = 0; return false" boilerplate.
+        //
+        public static bool TryParseThrowFormatException(out int bytesConsumed)
+        {
+            bytesConsumed = 0;
```
Comment (GrabYourPitchforks) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/ParserHelpers.cs]: Same nit as earlier.
